main.cc

```diff
@@ -146,6 +146,71 @@ read_config(bpo::variables_map& opts, db::config& cfg) {
         return make_exception_future<>(ep);
     });
 }
+
+// Handles SIGHUP, using it to trigger re-reading of the configuration file. Should
+// only be constructed on shard 0.
+class sighup_handler {
+    bpo::variables_map& _opts;
+    db::config& _cfg;
+    condition_variable _cond;
+    bool _pending = false; // if asked to reread while already reading
+    bool _stopping = false;
+    bool _stopped = false;
+    future<> _done = do_work();  // Launch main work loop, capture completion future
+public:
+    // Installs the signal handler. Must call stop() (and wait for it) before destruction.
```

Comment:
The downside is that it is impossible. Destructors execute atomically (in a single continuation), they cannot wait for a future (unless they're guaranteed to run in a thread).

See below how we use stop(), when we know we're in a thread, to implement RAII. The pattern is

    pseudo_raii_class_with_stop_or_close my_pseudo_raii_obj;
    auto stop_it = defer([&] { my_pseudo_raii_obj.stop().get(); });

We could sau that sigup_handler must run in a thread, but since the vast majority of the objects main starts are generic, it's better to conform to existing practice and keep sighup_handler similar to other stuff.